docs: add Web Component interop documentation (#30910)
microsoft/fluentui: apps/public-docsite-v9/src/Concepts/WebComponentsInterop/UsingFluentReactWithWebComponents.stories.mdx

import { Meta } from '@storybook/addon-docs';

<Meta title="Concepts/Developer/Web Components Interop/Using Fluent React with Web Components" />

> # Using Fluent React with Web Componenets

Reviewer: ```suggestion
# Using Fluent React with Web Components
```

Fluent React v9's extensible architecture and use of web platform features like CSS custom properties
allow it to be extended to work well with [Web Components](https://developer.mozilla.org/en-US/docs/Web/API/Web_components),
particularly [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_shadow_DOM).

> Where it is necessary to modify Fluent React's default behavior to better interoperate with Web Components, the
> implementation will live in the [Fluent UI Contrib repository](https://github.com/microsoft/fluentui-contrib).

Reviewer 2: I think this part is a little unclear, if you think of something better feel free to ignore my suggestion.
```suggestion
It is necessary to modify Fluent React v9's default behavior to better interoperate with Web Components. The
implementation lives in the [Fluent UI Contrib repository](https://github.com/microsoft/fluentui-contrib).
```

## Rendering Fluent React in shadow DOM

Fluent React components can be rendered inside shadow DOM with styles being set on the shadow root's `adoptedStyleSheets`
property.

```tsx
import { root } from '@fluentui-contrib/react-shadow';
import { FluentProvider, webLightTheme, Button } from '@fluentui/react-components';

<FluentProvider theme={webLightTheme}>
  {/* This is the shadow root */}
  <root.div>
    <Button>Fluent React Button in shadow DOM</Button>
  </root.div>
</FluentProvider>;
```

In the above example, note that `FluentProvider` sits outside the shadow DOM. When `FluentProvider` is inside the shadow
DOM styling/theming will not work as expected.

```tsx

> // ❌ This will not render correctly for example purposes only ❌

Reviewer 3: ```suggestion
// ❌ This will not render correctly, for example purposes only ❌
```

import { root } from '@fluentui-contrib/react-shadow';
import { FluentProvider, webLightTheme, Button } from '@fluentui/react-components';

/* This is the shadow root */
<root.div>
  <FluentProvider theme={webLightTheme}>
    <Button>Fluent React Button in shadow DOM</Button>
  </FluentProvider>
</root.div>;
```

To render a provider inside shadow DOM, use `ThemelessFluentProvider` instead.

> ## `ThemelessFluentProvider`

Reviewer 4: this PR should not be merged until this is released right ? https://github.com/microsoft/fluentui-contrib/pull/149

Author: Yep, as noted in the description 🙂

I moved it from draft for visibility since there was a fair bit of feedback on my recent documentation PRs.

Author: `ThemelessFluentProvider` is merged now. @layershifter when can we release it? I don't think I have access to the contrib release pipeline.

Author: Thanks @layershifter !

import { root } from '@fluentui-contrib/react-shadow';
import { createCSSStyleSheetFromTheme, ThemelessFluentProvider } from '@fluentui-contrib/react-themeless-provider';
import { webLightTheme, Button } from '@fluentui/react-components';

const themeSheet = createCSSStyleSheetFromTheme(':root', webLightTheme);
document.adoptedStyleSheets = [...document.adoptedStyleSheets, themeSheet];

const AppComponent = () => {
  // This must be called _before_ you render any Fluent React controls
  useShadowDOMSupport();
  <ThemelessFluentProvider>
    {/* This is the shadow root */}
    <root.div>
      <Button>First Button</Button>
      <Button>Second Button</Button>
    </root.div>
  </ThemelessFluentProvider>;
};
```

# Insertion Point

Fluent uses Griffel for authoring styles and it provides the [mergeClasses](https://griffel.js.org/react/api/merge-classes)
function for controlling style specificity. `mergeClasses` can accept any CSS class name but it only performs rule ordering
for class names generated by Griffel. This can lead to CSS specificity issues when you want to use CSS classes that were
not generated by Griffel (for example, application utility classes).

> Use the insertion point API to control the specificity of styles via document order.

Reviewer 3: _nit:_

```suggestion
For those cases, use the insertion point API to control the specificity of styles via document order.
```
